better handling of source errors

joesis opened the pull request:
For https://github.com/getlantern/lantern-internal/issues/3593

I should have requested for review when drafting the initial version of this repo, but well, the second best time is now.

joesis commented on keepcurrent.go:
```diff
@@ -40,67 +44,98 @@ type Runner struct {
+		var buf bytes.Buffer
+		if err == nil {
+			// Read ahead to surface any error reading from the source
+			_, err = io.Copy(&buf, rc)
```
I wonder if this can fix https://github.com/getlantern/lantern-internal/issues/3593#issuecomment-617167114. Previously, any error downloading the file is not treated as source error.

joesis replied:
Yeah makes sense, though it would make chaining sources or sinks harder. For example, `tarGzSource` reads the archive incrementally and returns once it finds the file expected, without requiring the whole archive being available.

joesis replied:
I should admit that part of the reason is that the `archiver` package https://github.com/mholt/archiver/blob/master/targz.go doesn't have an API to accept byte slice.

hwh33 replied:
Ah, good point about chaining, I didn't consider that.  An alternative could be to add a flag to the Runner to `SkipPreRead` (or something with a better name =P).  But that could just be added later if some user of the package desires.

hwh33 commented on keepcurrent.go:
```diff
@@ -40,67 +44,98 @@ type Runner struct {
 		}
 	}()
-	return func() { close(ch) }
+	return func() { close(chStop) }
```
This stop callback can return while `runner.syncOnce` continues to run.  Personally, I would be confused if this were to happen.  [Here's a PR](https://github.com/getlantern/keepcurrent/pull/2) that waits for `syncOnce` to return first.  It assumes good behavior of the source and sink though (they should stop eventually).  An alternative solution might be to check on `chStop` in `syncOnce` even after errors.  Currently, the channel is only checked if there was no error.

joesis replied:
Good catch. I'll just keep your fix.

joesis replied:
> Currently, the channel is only checked if there was no error.

The channel is checked when there was error and waited to retry, see https://github.com/getlantern/keepcurrent/pull/1/files#diff-f6f4ad39e1620037805a0b1c38a1e5eaR104

hwh33 replied:
Hmm, I misread that somehow.  Sounds good then!

hwh33 commented:
I think `webSource.Fetch` may be a bit fragile as well.  If a site does not support ETags or the "If-Modified-Since" header, it will always report new content.  You can actually see this in [my PR](https://github.com/getlantern/keepcurrent/pull/2) as the site used by `TestUpdateFromWeb` does not seem to respect either mechanism.

Perhaps `webSource.Fetch` could hash the content itself?

joesis commented:
> I think `webSource.Fetch` may be a bit fragile as well. If a site does not support ETags or the "If-Modified-Since" header, it will always report new content. 

IIUC the concern is the unnecessary write to the sinks when the file is identical to what they already have? I don't feel it much a problem. Keeping content hash would still require download the whole file to verify. In that case, flushing sinks would be more straightforward.

joesis commented:
The stalled test revealed a more serious bug in my code - `syncOnce` never returns if there's no error. Fixed at https://github.com/getlantern/keepcurrent/pull/1/commits/e36609e3fc7a1e872286680bba519bfca2535135 . Thanks for catching it @hwh33 ! Also added Travis.

hwh33 commented:
> Keeping content hash would still require download the whole file to verify.

Oh, good point.  Yeah, you're right then, it makes more sense to just update if we've already downloaded the file.

> The stalled test revealed a more serious bug in my code - syncOnce never returns if there's no error. Fixed at e36609e .

Nice!  Yes, much better now.

joesis commented:
Thanks @hwh33 for the review and hunting those hidden bugs!
